builtin/provisioners/habitat/linux_provisioner_test.go

```diff
@@ -0,0 +1,347 @@
+				"listen_http":        "192.168.20.1:8080",
+				"builder_auth_token": "dead-beef",
+				"gateway_auth_token": "ea7-beef",
+				"ctl_secret":         "bad-beef",
+			},
+
+			Commands: map[string]bool{
+				"env HAB_NONINTERACTIVE=true HAB_NOCOLORING=true HAB_AUTH_TOKEN=dead-beef sudo -E /bin/bash -c 'hab install core/hab-sup/0.79.1'":                             true,
+				"env HAB_NONINTERACTIVE=true HAB_NOCOLORING=true HAB_AUTH_TOKEN=dead-beef sudo -E /bin/bash -c 'systemctl enable hab-sup && systemctl start hab-sup'":         true,
+				"env HAB_NONINTERACTIVE=true HAB_NOCOLORING=true HAB_AUTH_TOKEN=dead-beef sudo -E /bin/bash -c 'mv /tmp/hab-sup.service /etc/systemd/system/hab-sup.service'": true,
+			},
+
+			Uploads: map[string]string{
+				"/tmp/hab-sup.service": linuxCustomSystemdUnitFileContents,
+			},
+		},
+	}
+
+	o := new(terraform.MockUIOutput)
+	c := new(communicator.MockCommunicator)
+
+	for k, tc := range cases {
+		c.Commands = tc.Commands
+		c.Uploads = tc.Uploads
+
+		p, err := decodeConfig(
+			schema.TestResourceDataRaw(t, Provisioner().(*schema.Provisioner).Schema, tc.Config),
+		)
+		if err != nil {
+			t.Fatalf("Error: %v", err)
+		}
+
+		err = p.linuxStartHabitat(o, c)
+		if err != nil {
+			t.Fatalf("Test %q failed: %v", k, err)
+		}
+	}
+}
+
+func TestLinuxProvisioner_linuxUploadRingKey(t *testing.T) {
+	cases := map[string]struct {
+		Config   map[string]interface{}
+		Commands map[string]bool
+	}{
+		"Upload ring key": {
+			Config: map[string]interface{}{
+				"version":          "0.79.1",
+				"auto_update":      true,
+				"use_sudo":         true,
+				"service_name":     "hab-sup",
+				"peers":            []string{"1.2.3.4"},
+				"ring_key":         "test-ring",
+				"ring_key_content": "dead-beef",
+			},
+
+			Commands: map[string]bool{
+				"env HAB_NONINTERACTIVE=true HAB_NOCOLORING=true sudo -E /bin/bash -c 'echo -e \"dead-beef\" | hab ring key import'": true,
+			},
+		},
+	}
+
+	o := new(terraform.MockUIOutput)
+	c := new(communicator.MockCommunicator)
+
+	for k, tc := range cases {
+		c.Commands = tc.Commands
+
+		p, err := decodeConfig(
+			schema.TestResourceDataRaw(t, Provisioner().(*schema.Provisioner).Schema, tc.Config),
+		)
+		if err != nil {
+			t.Fatalf("Error: %v", err)
+		}
+
+		err = p.linuxUploadRingKey(o, c)
+		if err != nil {
+			t.Fatalf("Test %q failed: %v", k, err)
+		}
+	}
+}
+
+func TestLinuxProvisioner_linuxStartHabitatService(t *testing.T) {
+	cases := map[string]struct {
+		Config   map[string]interface{}
+		Commands map[string]bool
+		Uploads  map[string]string
+	}{
+		"Start Habitat service with sudo": {
+			Config: map[string]interface{}{
+				"version":          "0.79.1",
+				"auto_update":      false,
+				"use_sudo":         true,
+				"service_name":     "hab-sup",
+				"peers":            []string{"1.2.3.4"},
+				"ring_key":         "test-ring",
+				"ring_key_content": "dead-beef",
+				"service": []map[string]interface{}{
```

Comment:
I am likely not explaining it in the best detail, but TF should generate a raw config of `[]interface{} { map[string]interface{}  } from an HCL config. In this testcase you have declared the datastructure as `[]map[string]interface{}` directly. You can change this to

```
...
"service": []interface{}{
    map[string]interface{}{
        ...
    }
},
```